src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Windows/Native/SafeHandles.cs

```diff
@@ -114,34 +114,40 @@ public static void DeleteKeyContainer(SafeCertContextHandle pCertContext)
                 {
                     CRYPT_KEY_PROV_INFO* pProvInfo = (CRYPT_KEY_PROV_INFO*)pProvInfoAsBytes;
 
-                    // For UWP the key was opened in a CNG-only context, so it can/should be deleted via CngKey.Delete.
-                    // In all other contexts, the key was loaded into CAPI, so deleting it via CryptAcquireContext is
-                    // the correct action.
-#if NETNATIVE
-                    string providerName = Marshal.PtrToStringUni((IntPtr)(pProvInfo->pwszProvName));
-                    string keyContainerName = Marshal.PtrToStringUni((IntPtr)(pProvInfo->pwszContainerName));
-
-                    try
+                    if (pProvInfo->dwProvType == 0)
```

Comment:
>  // For UWP the key was opened in a CNG-only context, so it can/should be deleted via CngKey.Delete.

>  // In all other contexts, the key was loaded into CAPI, so deleting it via CryptAcquireContext is

>  // the correct action.

How about we change this to:

// dwProvType being 0 indicates that the key is stored in a CNG handle.
// dwProvType being non-zero indicates that the key is stored in a CAPI handle.


From: Jeremy Barton [mailto:notifications@github.com]
Sent: Monday, April 18, 2016 8:01 AM
To: dotnet/corefx <corefx@noreply.github.com>
Cc: Atsushi Kanamori <atsushik@microsoft.com>
Subject: Re: [dotnet/corefx] Stop leaking keys on disk when PFX using CNG keys are opened. (#7828)


In src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Windows/Native/SafeHandles.cs<https://github.com/dotnet/corefx/pull/7828#discussion_r60073947>:

> @@ -114,34 +114,40 @@ public static void DeleteKeyContainer(SafeCertContextHandle pCertContext)

>                  {

>                      CRYPT_KEY_PROV_INFO* pProvInfo = (CRYPT_KEY_PROV_INFO*)pProvInfoAsBytes;

>

> -                    // For UWP the key was opened in a CNG-only context, so it can/should be deleted via CngKey.Delete.

> -                    // In all other contexts, the key was loaded into CAPI, so deleting it via CryptAcquireContext is

> -                    // the correct action.

> -#if NETNATIVE

> -                    string providerName = Marshal.PtrToStringUni((IntPtr)(pProvInfo->pwszProvName));

> -                    string keyContainerName = Marshal.PtrToStringUni((IntPtr)(pProvInfo->pwszContainerName));

> -

> -                    try

> +                    if (pProvInfo->dwProvType == 0)

This is probably worth a comment along the lines of "When opening the key the dwProvType==0 check isn't correct, because most CAPI keys can be loaded in CNG. But for deleting the key handle this check is correct, because it's asking about the storage of the key, not the functionality of it".

Because I was in the middle of asking why it was valid here. (The comment in the if doesn't quite get it, since it talks about how the key was opened)

—
You are receiving this because you authored the thread.
Reply to this email directly or view it on GitHub<https://github.com/dotnet/corefx/pull/7828/files/32044d812753b880a65aad778b5ec362fcfcb16a#r60073947>